src/python/pants/backend/python/subsystems/setup.py

```diff
@@ -226,13 +232,22 @@ class PythonSetup(Subsystem):
             Expects a dictionary of resolve names from `[python].resolves` and Python tools (e.g.
             `black` and `pytest`) to file paths for
             constraints files. For example,
-            `{'data-science': '3rdparty/data-science-constraints.txt'}`.
-            If a resolve is not set in the dictionary, it will not use a constraints file.
+            `{{'data-science': '3rdparty/data-science-constraints.txt'}}`. If a resolve is not set
+            in the dictionary, it will not use a constraints file.
 
-            You can use the key `__default__` to set a default value for all resolves.
+            You can use the key `{RESOLVE_OPTION_KEY__DEFAULT}` to set a default value for all
+            resolves.
 
-            Note: Only takes effect if you use Pex lockfiles. Use the default
-            `[python].lockfile_generator = "pex"` and run the `generate-lockfiles` goal.
+            The constraints file will still be used if you are using a custom-managed lockfile
+            for a particular resolve, or if you disabled lockfiles for a certain tool. If you are
+            not using `[python].resolves` yet, use the key `{RESOLVE_OPTION_KEY__NO_USER_RESOLVE}`
```

Comment:
> I though the plan was to eventually mandate locks.

This has not been finalized. We require it for JVM, but have not committed to that for Python.

> If so, this is adding a feature that will go away, in which case why not just not add the feature. If you want to use constraints in the classic Pip use-case way you must use lock files.

So, the more important options are `[python].resolves_to_find_links` and `[python].resolves_to_indexes`. In the [design doc's deprecation plan](https://docs.google.com/document/d/1HAvpSNvNAHreFfvTAXavZGka-A3WWvPuH0sMjGUCo48/edit#heading=h.2xanhw6bnjq0), I propose migrating `[python-repos]` into those new `[python]` options, rather than keeping around the old and new, which IMO is confusing.

As explained there, I think it's desirable to let you change the options for `no-user-resolve` while letting other resolves have other options. My first design was that you must use `__default__` to set `no-user-resolve`, until this new improved idea.

So, if we have `no-user-resolve` already for `[python].resolves_to_find_links`, it is trivial to also support constraints files.

Reply:
Was there ever any feedback from the 1 user that said they'd like this feature (`[python].resolves_to_find_links` + `[python].resolves_to_indexes`)? IIRC they actually just said only having a global setting for each of these "was a pain".

Reply:
That user did not reply.

I revised the design doc's motivation section:

> In the past, we've had issues with [python].requirement_constraints applying to [tool lockfiles too](https://github.com/pantsbuild/pants/pull/12639) because sometimes users needed different pins for tools than user code, so we reverted that behavior. Generally, the multiple resolves feature is intended to give users flexibility, such as using two different versions of Django in the same repository; likewise, it's reasonable to need two different versions of constraints in a repo. This proposal builds off of that flexibility to also allow customizing settings like when to use a custom index. Most users won't need this flexibility, but we can expect some will. For example, a user today said:
> 
> "Our custom repo is somewhat problematic, so ability to namespace it and only reference it when needed would probably be a win."

Reply:
I don't think the new words change the situation. I still see no actual reason a non-global configuration is needed for repos + indexes. It would be great to have a use case with no reasonable workaround before basing reasoning off it.

Reply:
I don't want to strongarm us migrating from `[python-repos]` to `[python].resolves_to_{indexes,find_links}`, nor migrating `[python].manylinux` to be per-resolve. So I will revert the parts of this PR that are prework for those changes, and have this PR solely make sense in the context of per-resolve constraints files.

Then, we can evaluate whether to migrate each option individually.

I can also revert `no-user-resolve` and leave it for the `[python].requirements_constraint` deprecation, so it can be considered in one PR.

(Seems like there is little controversy that per-resolve constraints files make sense, given we allow the input requirements themselves to differ per resolve.)